File: esteira/src/garimpo_esteira/state_machine.py
"""Máquina de estados — espelha lead_status_transitions (Fase 0).

O banco é a fonte da verdade (trigger valida). Aqui replicamos só o que a
esteira precisa: avançar bruto -> enriquecido (e descartar quando for o caso).
"""
from __future__ import annotations

from .models import LeadStatus

TRANSITIONS: dict[LeadStatus, tuple[LeadStatus, ...]] = {
    "bruto": ("enriquecido", "descartado"),
    "enriquecido": ("qualificado", "descartado"),
    "qualificado": ("rascunho_pronto", "descartado"),
    "rascunho_pronto": ("aprovado", "descartado"),
    "aprovado": ("enviado",),
    "enviado": ("respondeu", "sem_resposta", "descartado"),
    "sem_resposta": ("enviado", "descartado"),
    "respondeu": ("interessado", "sem_interesse", "reuniao"),
    "interessado": ("reuniao", "proposta", "perdido"),
    "reuniao": ("proposta", "perdido"),
    "proposta": ("fechado", "perdido"),
    "descartado": ("enriquecido",),  # reativavel pelo dono (volta pro funil)
    "sem_interesse": (),
    "fechado": (),
    "perdido": (),
}

CONTACT_STATUSES: frozenset[LeadStatus] = frozenset({"rascunho_pronto", "aprovado", "enviado"})


def can_transition(src: LeadStatus, dst: LeadStatus, opt_out: bool = False) -> bool:
    if dst not in TRANSITIONS.get(src, ()):
        return False
    if opt_out and dst in CONTACT_STATUSES:
        return False
    return True

File: extension/src/lib/state-machine.mjs
// Maquina de estados — espelha lead_status_transitions (Fase 0).
// Os botoes contextuais da extensao (secao 6 do mapa) saem daqui.

export const TRANSITIONS = {
  bruto: ["enriquecido", "descartado"],
  enriquecido: ["qualificado", "descartado"],
  qualificado: ["rascunho_pronto", "descartado"],
  rascunho_pronto: ["aprovado", "descartado"],
  aprovado: ["enviado"],
  enviado: ["respondeu", "sem_resposta", "descartado"],
  sem_resposta: ["enviado", "descartado"],
  respondeu: ["interessado", "sem_interesse", "reuniao"],
  interessado: ["reuniao", "proposta", "perdido"],
  reuniao: ["proposta", "perdido"],
  proposta: ["fechado", "perdido"],
  descartado: [],
  sem_interesse: [],
  fechado: [],
  perdido: [],
};

export const STATUS_LABEL = {
  bruto: "Bruto",
  enriquecido: "Enriquecido",
  qualificado: "Qualificado",
  rascunho_pronto: "Rascunho pronto",
  aprovado: "Aprovado",
  enviado: "Enviado",
  sem_resposta: "Sem resposta",
  respondeu: "Respondeu",
  interessado: "Interessado",
  reuniao: "Reuniao",
  proposta: "Proposta",
  fechado: "Fechado",
  descartado: "Descartado",
  sem_interesse: "Sem interesse",
  perdido: "Perdido",
};

// Rotulos dos botoes (iguais aos do front/mapa). Default = STATUS_LABEL[to].
const TRANSITION_LABELS = {
  "enviado->descartado": "Numero errado",
  "respondeu->reuniao": "Agendou reuniao",
  "rascunho_pronto->aprovado": "Aprovar",
  "aprovado->enviado": "Marquei enviado",
  "sem_resposta->enviado": "Reenviei (follow-up)",
};

const CONTACT_STATUSES = new Set(["rascunho_pronto", "aprovado", "enviado"]);

export function transitionLabel(from, to) {
  return TRANSITION_LABELS[`${from}->${to}`] || STATUS_LABEL[to];
}

// Botoes contextuais para o status atual: [{ to, label, blocked }]
export function contextualButtons(status, optOut = false) {
  return (TRANSITIONS[status] || []).map((to) => ({
    to,
    label: transitionLabel(status, to),
    blocked: optOut && CONTACT_STATUSES.has(to),
  }));
}
